### ISSUE TITLE
Inconsistent delegate behavior...

### ISSUE THREAD

**Issue by jinks** [KSP-KOS/KOS#2167]
I'm currently working my way through the babylonic tower of indirection that is @gisikw's knu and mission runner system.

In the process I discovered that the latest version (ep045) doesn't work with events at all.

The problematic code is `for v in e:values v().` inside his `mission.ks`. At that point `e` is `lex("fairing", {print "Event f". if alt:radar > 50000 {stage. ev:remove("fairing").})` in my mission.

I get a "too few arguments" error, which was extremely hard to track down, since, obviously, that delegate doesn't take any arguments.

After going through most of the issue tracker, I'm pretty sure this is the same issue as #1727 and kOS is trying to construct a vector here instead of calling `e["fairing"]`<sup>1</sup>, but that one hasn't seen any comment in one and a half years.

Unlike #1727 though, changing `v()` to `v:call()` resolves the issue.

I'll attach the relevant log in the hope that it'll be helpful for someone:
[kos-delegate.log](https://github.com/KSP-KOS/KOS/files/1465773/kos-delegate.log)

---

<sup>1</sup>: The fact that kOS expects 3 arguments and wants them to be scalar supports this notion. :) 
As a side note, it would be nice if the error was more specific. Something along the lines of "Expected x, got y arguments".

**Comment by Tricky1702** [comment]
You are missing a closing }

Try this

  lex("fairing", {print "Event f". if alt:radar > 50000 {stage.
ev:remove("fairing").}})


On 13 Nov 2017 06:22, "Jinks" <notifications@github.com> wrote:

I'm currently working my way through the babylonic tower of indirection
that is @gisikw <https://github.com/gisikw>'s knu and mission runner system.

In the process I discovered that the latest version (ep045) doesn't work
with events at all.

The problematic code is for v in e:values v(). inside his mission.ks. At
that point e is lex("fairing", {print "Event f". if alt:radar > 50000
{stage. ev:remove("fairing").}) in my mission.

I get a "too few arguments" error, which was extremely hard to track down,
since, obviously, that delegate doesn't take any arguments.

After going through most of the issue tracker, I'm pretty sure this is the
same issue as #1727 <https://github.com/KSP-KOS/KOS/issues/1727> and kOS is
trying to construct a vector here instead of calling e["fairing"]1, but
that one hasn't seen any comment in one and a half years.

Unlike #1727 <https://github.com/KSP-KOS/KOS/issues/1727> though, changing
v() to v:call() resolves the issue.

I'll attach the relevant log in the hope that it'll be helpful for someone:
kos-delegate.log
<https://github.com/KSP-KOS/KOS/files/1465773/kos-delegate.log>
------------------------------

1: The fact that kOS expects 3 arguments and wants them to be scalar
supports this notion. :)
As a side note, it would be nice if the error was more specific. Something
along the lines of "Expected x, got y arguments".

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/KSP-KOS/KOS/issues/2167>, or mute the thread

**Comment by jinks** [comment]
> You are missing a closing }


Copy & paste error. It **is** running fine, it's just running into a limitation of kOS in regard to variable naming.

**Comment by hvacengi** [comment]
Because built in functions are handled differently from user functions (the underlying opcode treats these two differently) kOS doesn't really support having user functions that match built in function names.  This is because built in functions are handled when the program is compiled, and user functions are handled when they are actually called/referenced.  Under the current conditions we need two things:

- A note in the documentation that variable/function names cannot match a built in function name
- An error thrown when attempting to declare a conflicting function and/or variable.

**Comment by jinks** [comment]
@hvacengi It's probably too late to change it now, buy maybe it was a bad idea for kOS to hog single-letter function and variable names, especially `v()` and `r()` which often get used for things like velocity or radius instead of vector and rotation.

(The fact that variables and functions kinda live in the same namespace doesn't really help matters here either.)

**Comment by Dunbaratu** [comment]
@jinks you are correct that it was a bad idea.  But that bad idea pre-dates all the current people working on the project.  It was a decision made in the old original kOS before we touched it, and we had the exact same problem: "probably too late to change it now".  Not from our end, but from the fact that there was already a userbase to think about:

For every person saying what you're saying, we'd have just as many people saying, "I upgraded kOS and those idiots broke all my scripts, what?  Don't they understand that backward compatibility is really important?  Don't they know anything about supporting a living system that's actually being used?"  There's no way to fix it without breaking old scripts - script examples people post to out on fora and websites that we have no control over, and show up in google searches.  At best the most we can do is try to do a better job of making it clear what happened when it doesn't work.

**Comment by jinks** [comment]
@Dunbaratu Yeah, that's what I mean. In the long run _every_ programming language runs into that problem and, to my knowledge, nobody has found a satisfactory way to deal with it.

C++ just tacks on new stuff on top of the old and tells people to please stop using the old way but they're still there.

Python and Go solve it via "We reserve the right to make breaking changes between major versions". It works for them because they can rely on a mostly static environment, you can just run your old scripts with the old compiler because, unlike KSP, OSs go to great lengths to maintain backwards compatibility so it's easy to keep old versions around. (Kinda hard to do when you can't even be sure that _physics_ is gonna work the same way a year down the road.)

Rust opted for an "epoch" system, where some metadata in the user code decides which language version to apply (with "no metadata" meaning "epoch 0"). This of course means there are several internal code paths to maintain that resolve wildly different source code to the same opcodes.

For kOS, if you even ever **want** to majorly change the language, I think a mix of 2 and 3 is the most viable. Keep the VM as-is but have a new language on top of that which resolves to the same opcodes. Make it clear that "KerboScript++" is incompatible with old versions. The extreme version of that would be to have pluggable compilers for a multitude of languages, i.e. kOS would take on a form similar to the JVM or CLR. Sounds like a ton of work, though.
